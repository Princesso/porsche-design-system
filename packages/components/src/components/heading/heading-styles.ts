import type { JssStyle } from 'jss';
import type { HeadingVariant, TextAlign, TextColor, TextSize, Theme, VariantType } from '../../types';
import { buildResponsiveStyles, getCss, textMap } from '../../utils';
import { addImportantToEachRule, getThemedColors } from '../../styles';
import {
  headingLargeStyle,
  headingXLargeStyle,
  headingMediumStyle,
  headingSmallStyle,
  headingXXLargeStyle,
  fontWeight,
  textSmallStyle,
  displayLargeStyle,
} from '@porsche-design-system/utilities-v2';
import { getEllipsisJssStyle, getSlottedTypographyJssStyle } from '../../styles/typography-styles';
import { isVariantType } from './heading-utils';

const headingMap: Record<VariantType, JssStyle> = {
  'large-title': displayLargeStyle,
  'heading-1': headingXXLargeStyle,
  'heading-2': headingXLargeStyle,
  'heading-3': headingLargeStyle,
  'heading-4': headingMediumStyle,
  'heading-5': headingSmallStyle,
  'headline-1': headingXXLargeStyle, // deprecated
  'headline-2': headingXLargeStyle, // deprecated
  'headline-3': headingLargeStyle, // deprecated
  'headline-4': headingMediumStyle, // deprecated
  'headline-5': headingSmallStyle, // deprecated
};

const getVariantJssStyle = (variant: HeadingVariant): JssStyle => {
  return headingMap[variant as VariantType];
};

const getSizeJssStyle = (textSize: TextSize): JssStyle => {
  const { semiBold: fontWeightSemiBold } = fontWeight;
  return textSize === 'inherit'
    ? {
        fontSize: textSize,
        fontWeight: fontWeightSemiBold,
      }
    : {
        font: textMap[textSize].font.replace('400', fontWeightSemiBold),
      };
};

export const getComponentCss = (
  variant: HeadingVariant,
  align: TextAlign,
  color: Extract<TextColor, 'primary' | 'default' | 'inherit'>,
  ellipsis: boolean,
  theme: Theme
): string => {
  return getHeadingHeadlineStyles(variant, align, color, ellipsis, theme);
};

export const getHeadingHeadlineStyles = (
  variant: HeadingVariant,
  align: TextAlign,
  color: Extract<TextColor, 'primary' | 'default' | 'inherit'>,
  ellipsis: boolean,
  theme: Theme
): string => {
  return getCss({
    '@global': {
      ':host': {
        display: 'block',
      },
      '::slotted': {
        '&(h1),&(h2),&(h3),&(h4),&(h5),&(h6)': addImportantToEachRule(getSlottedTypographyJssStyle()),
      },
    },
    root: {
      padding: 0,
      margin: 0,
      textAlign: align,
      color: color === 'inherit' ? 'inherit' : getThemedColors(theme).primaryColor,
      whiteSpace: 'inherit',
      ...(isVariantType(variant)
        ? getVariantJssStyle(variant)
        : {
            ...textSmallStyle,
            ...buildResponsiveStyles(variant, getSizeJssStyle),
            overflowWrap: null,
            hyphens: null,
          }),
      ...(ellipsis && getEllipsisJssStyle()),
    },
  });
};
