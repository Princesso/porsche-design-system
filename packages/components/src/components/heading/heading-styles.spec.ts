import { getComponentCss } from './heading-styles';

describe('getComponentCss()', () => {
  it.each<Parameters<typeof getComponentCss>>([
    ['heading-1', 'left', 'default', false, 'light'],
    ['inherit', 'left', 'default', false, 'light'],
    ['large-title', 'center', 'inherit', true, 'dark'],
    ['heading-2', 'center', 'inherit', true, 'dark'],
    ['heading-3', 'center', 'inherit', true, 'dark'],
    ['heading-4', 'center', 'inherit', true, 'dark'],
    ['heading-5', 'center', 'inherit', true, 'dark'],
    [
      { base: 'small', xs: 'large', s: 'small', m: 'large', l: 'small', xl: 'large' },
      'right',
      'inherit',
      false,
      'dark',
    ],
  ])('should return correct css for variant: %j, align: %s, color: %s, ellipsis: %s and theme: %s', (...args) => {
    expect(getComponentCss(...args)).toMatchSnapshot();
  });
});
