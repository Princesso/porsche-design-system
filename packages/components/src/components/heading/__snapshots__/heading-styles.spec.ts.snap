// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getComponentCss() should return correct css for variant: "heading-1", align: left, color: default, ellipsis: false and theme: light 1`] = `
":host {
  display: block;
}
:host([hidden]) {
  display: none !important;
}
::slotted(h1), ::slotted(h2), ::slotted(h3), ::slotted(h4), ::slotted(h5), ::slotted(h6) {
  margin: inherit !important;
  padding: inherit !important;
  font-family: inherit !important;
  font-weight: inherit !important;
  font-size: inherit !important;
  line-height: inherit !important;
  font-style: inherit !important;
  font-variant: inherit !important;
  color: inherit !important;
  text-align: inherit !important;
  overflow-wrap: inherit !important;
  word-wrap: inherit !important;
  hyphens: inherit !important;
  white-space: inherit !important;
}
.root {
  padding: 0;
  margin: 0;
  text-align: left;
  color: #010205;
  white-space: inherit;
  font: normal normal 600 clamp(2.07rem, .96vw + 1.88rem, 2.94rem)/calc(6px + 2.125ex) 'Porsche Next','Arial Narrow',Arial,'Heiti SC',SimHei,sans-serif;
}"
`;

exports[`getComponentCss() should return correct css for variant: "heading-2", align: center, color: inherit, ellipsis: true and theme: dark 1`] = `
":host {
  display: block;
}
:host([hidden]) {
  display: none !important;
}
::slotted(h1), ::slotted(h2), ::slotted(h3), ::slotted(h4), ::slotted(h5), ::slotted(h6) {
  margin: inherit !important;
  padding: inherit !important;
  font-family: inherit !important;
  font-weight: inherit !important;
  font-size: inherit !important;
  line-height: inherit !important;
  font-style: inherit !important;
  font-variant: inherit !important;
  color: inherit !important;
  text-align: inherit !important;
  overflow-wrap: inherit !important;
  word-wrap: inherit !important;
  hyphens: inherit !important;
  white-space: inherit !important;
}
.root {
  padding: 0;
  margin: 0;
  text-align: center;
  color: inherit;
  white-space: nowrap;
  font: normal normal 600 clamp(1.73rem, .57vw + 1.61rem, 2.24rem)/calc(6px + 2.125ex) 'Porsche Next','Arial Narrow',Arial,'Heiti SC',SimHei,sans-serif;
  max-width: 100%;
  overflow: hidden;
  text-overflow: ellipsis;
}"
`;

exports[`getComponentCss() should return correct css for variant: "heading-3", align: center, color: inherit, ellipsis: true and theme: dark 1`] = `
":host {
  display: block;
}
:host([hidden]) {
  display: none !important;
}
::slotted(h1), ::slotted(h2), ::slotted(h3), ::slotted(h4), ::slotted(h5), ::slotted(h6) {
  margin: inherit !important;
  padding: inherit !important;
  font-family: inherit !important;
  font-weight: inherit !important;
  font-size: inherit !important;
  line-height: inherit !important;
  font-style: inherit !important;
  font-variant: inherit !important;
  color: inherit !important;
  text-align: inherit !important;
  overflow-wrap: inherit !important;
  word-wrap: inherit !important;
  hyphens: inherit !important;
  white-space: inherit !important;
}
.root {
  padding: 0;
  margin: 0;
  text-align: center;
  color: inherit;
  white-space: nowrap;
  font: normal normal 600 clamp(1.44rem, .3vw + 1.38rem, 1.71rem)/calc(6px + 2.125ex) 'Porsche Next','Arial Narrow',Arial,'Heiti SC',SimHei,sans-serif;
  max-width: 100%;
  overflow: hidden;
  text-overflow: ellipsis;
}"
`;

exports[`getComponentCss() should return correct css for variant: "heading-4", align: center, color: inherit, ellipsis: true and theme: dark 1`] = `
":host {
  display: block;
}
:host([hidden]) {
  display: none !important;
}
::slotted(h1), ::slotted(h2), ::slotted(h3), ::slotted(h4), ::slotted(h5), ::slotted(h6) {
  margin: inherit !important;
  padding: inherit !important;
  font-family: inherit !important;
  font-weight: inherit !important;
  font-size: inherit !important;
  line-height: inherit !important;
  font-style: inherit !important;
  font-variant: inherit !important;
  color: inherit !important;
  text-align: inherit !important;
  overflow-wrap: inherit !important;
  word-wrap: inherit !important;
  hyphens: inherit !important;
  white-space: inherit !important;
}
.root {
  padding: 0;
  margin: 0;
  text-align: center;
  color: inherit;
  white-space: nowrap;
  font: normal normal 600 clamp(1.2rem, .12vw + 1.18rem, 1.31rem)/calc(6px + 2.125ex) 'Porsche Next','Arial Narrow',Arial,'Heiti SC',SimHei,sans-serif;
  max-width: 100%;
  overflow: hidden;
  text-overflow: ellipsis;
}"
`;

exports[`getComponentCss() should return correct css for variant: "heading-5", align: center, color: inherit, ellipsis: true and theme: dark 1`] = `
":host {
  display: block;
}
:host([hidden]) {
  display: none !important;
}
::slotted(h1), ::slotted(h2), ::slotted(h3), ::slotted(h4), ::slotted(h5), ::slotted(h6) {
  margin: inherit !important;
  padding: inherit !important;
  font-family: inherit !important;
  font-weight: inherit !important;
  font-size: inherit !important;
  line-height: inherit !important;
  font-style: inherit !important;
  font-variant: inherit !important;
  color: inherit !important;
  text-align: inherit !important;
  overflow-wrap: inherit !important;
  word-wrap: inherit !important;
  hyphens: inherit !important;
  white-space: inherit !important;
}
.root {
  padding: 0;
  margin: 0;
  text-align: center;
  color: inherit;
  white-space: nowrap;
  font: normal normal 600 1rem/calc(6px + 2.125ex) 'Porsche Next','Arial Narrow',Arial,'Heiti SC',SimHei,sans-serif;
  max-width: 100%;
  overflow: hidden;
  text-overflow: ellipsis;
}"
`;

exports[`getComponentCss() should return correct css for variant: "inherit", align: left, color: default, ellipsis: false and theme: light 1`] = `
":host {
  display: block;
}
:host([hidden]) {
  display: none !important;
}
::slotted(h1), ::slotted(h2), ::slotted(h3), ::slotted(h4), ::slotted(h5), ::slotted(h6) {
  margin: inherit !important;
  padding: inherit !important;
  font-family: inherit !important;
  font-weight: inherit !important;
  font-size: inherit !important;
  line-height: inherit !important;
  font-style: inherit !important;
  font-variant: inherit !important;
  color: inherit !important;
  text-align: inherit !important;
  overflow-wrap: inherit !important;
  word-wrap: inherit !important;
  hyphens: inherit !important;
  white-space: inherit !important;
}
.root {
  padding: 0;
  margin: 0;
  text-align: left;
  color: #010205;
  white-space: inherit;
  font: normal normal 400 1rem/calc(6px + 2.125ex) 'Porsche Next','Arial Narrow',Arial,'Heiti SC',SimHei,sans-serif;
  font-size: inherit;
  font-weight: 600;
}"
`;

exports[`getComponentCss() should return correct css for variant: "large-title", align: center, color: inherit, ellipsis: true and theme: dark 1`] = `
":host {
  display: block;
}
:host([hidden]) {
  display: none !important;
}
::slotted(h1), ::slotted(h2), ::slotted(h3), ::slotted(h4), ::slotted(h5), ::slotted(h6) {
  margin: inherit !important;
  padding: inherit !important;
  font-family: inherit !important;
  font-weight: inherit !important;
  font-size: inherit !important;
  line-height: inherit !important;
  font-style: inherit !important;
  font-variant: inherit !important;
  color: inherit !important;
  text-align: inherit !important;
  overflow-wrap: inherit !important;
  word-wrap: inherit !important;
  hyphens: inherit !important;
  white-space: inherit !important;
}
.root {
  padding: 0;
  margin: 0;
  text-align: center;
  color: inherit;
  white-space: nowrap;
  font: italic normal 600 clamp(2.49rem, 5.67vw + 1.35rem, 7.59rem)/calc(6px + 2.125ex) 'Porsche Next','Arial Narrow',Arial,'Heiti SC',SimHei,sans-serif;
  max-width: 100%;
  overflow: hidden;
  text-overflow: ellipsis;
}"
`;

exports[`getComponentCss() should return correct css for variant: {"base":"small","xs":"large","s":"small","m":"large","l":"small","xl":"large"}, align: right, color: inherit, ellipsis: false and theme: dark 1`] = `
":host {
  display: block;
}
:host([hidden]) {
  display: none !important;
}
::slotted(h1), ::slotted(h2), ::slotted(h3), ::slotted(h4), ::slotted(h5), ::slotted(h6) {
  margin: inherit !important;
  padding: inherit !important;
  font-family: inherit !important;
  font-weight: inherit !important;
  font-size: inherit !important;
  line-height: inherit !important;
  font-style: inherit !important;
  font-variant: inherit !important;
  color: inherit !important;
  text-align: inherit !important;
  overflow-wrap: inherit !important;
  word-wrap: inherit !important;
  hyphens: inherit !important;
  white-space: inherit !important;
}
.root {
  padding: 0;
  margin: 0;
  text-align: right;
  color: inherit;
  white-space: inherit;
  font: normal normal 600 1rem/calc(6px + 2.125ex) 'Porsche Next','Arial Narrow',Arial,'Heiti SC',SimHei,sans-serif;
}
@media(min-width:480px) {
  .root {
    font: normal normal 600 clamp(1.44rem, .3vw + 1.38rem, 1.71rem)/calc(6px + 2.125ex) 'Porsche Next','Arial Narrow',Arial,'Heiti SC',SimHei,sans-serif;
  }
}

@media(min-width:760px) {
  .root {
    font: normal normal 600 1rem/calc(6px + 2.125ex) 'Porsche Next','Arial Narrow',Arial,'Heiti SC',SimHei,sans-serif;
  }
}

@media(min-width:1000px) {
  .root {
    font: normal normal 600 clamp(1.44rem, .3vw + 1.38rem, 1.71rem)/calc(6px + 2.125ex) 'Porsche Next','Arial Narrow',Arial,'Heiti SC',SimHei,sans-serif;
  }
}

@media(min-width:1300px) {
  .root {
    font: normal normal 600 1rem/calc(6px + 2.125ex) 'Porsche Next','Arial Narrow',Arial,'Heiti SC',SimHei,sans-serif;
  }
}

@media(min-width:1760px) {
  .root {
    font: normal normal 600 clamp(1.44rem, .3vw + 1.38rem, 1.71rem)/calc(6px + 2.125ex) 'Porsche Next','Arial Narrow',Arial,'Heiti SC',SimHei,sans-serif;
  }
}
"
`;
